### PR TITLE
Auto-detect blob type as page blob for VHD

### PR DESCRIPTION
**Pull request by adreed-msft**
Turns out most of the infrastructure was already there, just had to shift copy and sync over to using it.

Should we auto-detect on sync as well?

### CODE REVIEW

File: testSuite/scripts/test_autodetect_blob_type.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+                                                                                                        "info"). \
+            add_flags("block-size-mb", 4).execute_azcopy_copy_command()
+        self.assertTrue(result)
+
+        # execute validator. Validator will ensure it's a page blob as validator now checks blob type before testing.
+        result = util.Command("testBlob").add_arguments(file_path).add_arguments(destination_sas). \
+            add_flags("blob-type", "PageBlob").execute_azcopy_verify()
+        self.assertTrue(result)
+
+    def test_copy_infer_blob_type_from_block_to_page_blob(self):
```
[the remark this answers is not in the document]
Comment (JohnRusk) [reply testSuite/scripts/test_autodetect_blob_type.py]: Also, @zezha-msft , should we infer when copying from Azure Files?  Whatever we do, should be consistent.  I think the three possible ways to do that are:
(a) Always infer (even when source is Blob Storage) (this would cover all sources)  OR
(b) Always infer when source does not have a blob type (i.e. when source is not Blob Storage). This covers AWS and Azure files as sources, plus local disk., OR
(c) Infer only when uploading from local disk.

Right now we have (c) and part of (a).

File: ste/xfer.go [hunk 2]
```diff
@@ -76,15 +76,7 @@ func computeJobXfer(fromTo common.FromTo, blobType common.BlobType) newJobXfer {
...
-		switch blobType {
-		case common.EBlobType.None(),
-			common.EBlobType.BlockBlob():
-			return parameterizeSend(anyToRemote, newBlockBlobUploader, newLocalSourceInfoProvider)
-		case common.EBlobType.PageBlob():
-			return parameterizeSend(anyToRemote, newPageBlobUploader, newLocalSourceInfoProvider)
-		case common.EBlobType.AppendBlob():
-			return parameterizeSend(anyToRemote, newAppendBlobUploader, newLocalSourceInfoProvider)
-		}
+		return parameterizeSend(anyToRemote, newBlobUploader, newLocalSourceInfoProvider)
```
Comment (JohnRusk) [inline ste/xfer.go]: I haven't tested it to check, but it _looks_ like this change has broken the ability to specify a specific blob type on the command line, for uploads from local disk. (Because it ignores the blobType param that is passed in to computeJobXfer).  Can you test to check whether it has indeed broken that ability?

If it has, then we should consider a few things:
1. Obviously we need it to work
2. But should it be made to work the way it did here... OR
3. Should we align it more closely to the way the same command-line parameter data is used for S2S copying (which respects the command line param, but by a different mechanism in the form of `jptm.BlobTypeOverride()` . That routine ultimately reads the same command line-supplied value data , namely `plan.DstBlobData.BlobType`)

I'll put a suggestion about this in my overall comment on the PR.

File: testSuite/scripts/test_autodetect_blob_type.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JohnRusk) [reply testSuite/scripts/test_autodetect_blob_type.py]: We'll automatically get (b) if we follow the approach I suggest in my overall end-of-review comment (to introduce one getTargetBlobType method)

File: cmd/copyS2SMigrationBlobEnumerator.go [hunk 3]
```diff
@@ -260,6 +266,13 @@ func (e *copyS2SMigrationBlobEnumerator) addBlobToNTransfer(srcURL, destURL url.
 
 func (e *copyS2SMigrationBlobEnumerator) addBlobToNTransfer2(srcURL, destURL url.URL, properties *azblob.BlobGetPropertiesResponse,
 	cca *cookedCopyCmdArgs) error {
+	util := copyHandlerUtil{}
+
+	bt := properties.BlobType()
+	if e.BlobAttributes.BlobType == common.EBlobType.None() || e.BlobAttributes.BlobType == common.EBlobType.Detect() {
+		bt = util.inferBlobType(srcURL.Path, bt)
```
[the remark this answers is not in the document]
Comment (JohnRusk) [reply cmd/copyS2SMigrationBlobEnumerator.go]: Note that there's not need to specifically address this comment, other than running the test, if you do the `getTargetBlobType` approach that I suggested in my overall end-of-review comment.

File: cmd/copyS2SMigrationBlobEnumerator.go [hunk 4]
```diff
@@ -239,6 +239,12 @@ func (e *copyS2SMigrationBlobEnumerator) addTransfersFromContainer(ctx context.C
 
 func (e *copyS2SMigrationBlobEnumerator) addBlobToNTransfer(srcURL, destURL url.URL, properties *azblob.BlobProperties, metadata azblob.Metadata,
 	cca *cookedCopyCmdArgs) error {
+	util := copyHandlerUtil{}
```
[the remark this answers is not in the document]
Comment (JohnRusk) [reply cmd/copyS2SMigrationBlobEnumerator.go]: No need to address this comment (other than to revert this code change to the prior-to-PR state) if you use the `getTargetBlobType` approach suggested in my overal-end-of-review comment

**Comment by JohnRusk** [comment]
Note that moving the inference into the STE with `getTargetBlobType` would cause it to apply to sync as well. cc @zezha-msft  in case he thinks we don't want it to apply to sync.

File: testSuite/scripts/test_autodetect_blob_type.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (adreed-msft) [reply testSuite/scripts/test_autodetect_blob_type.py]: I think this would be a good blanket solution:

Infer blob type by default for sources that do not have a blob type, and only infer when explicitly told to from sources with a blob type.

File: cmd/copyUtil.go [hunk 5]
```diff
@@ -49,6 +49,21 @@ var gCopyUtil = copyHandlerUtil{}
 
 const wildCard = "*"
 
+var inferExtensions = map[string]azblob.BlobType{
+	".vhd":  azblob.BlobPageBlob,
+	".vhdx": azblob.BlobPageBlob,
+	".vmdk": azblob.BlobPageBlob,
```
[the remark this answers is not in the document]
Comment (adreed-msft) [reply cmd/copyUtil.go]: Ah, I was just blanket covering the virtual disks I knew of.

File: cmd/copy.go [hunk 6]
```diff
@@ -1163,7 +1163,7 @@ func init() {
 	// options change how the transfers are performed
 	cpCmd.PersistentFlags().Float64Var(&raw.blockSizeMB, "block-size-mb", 0, "use this block size (specified in MiB) when uploading to/downloading from Azure Storage. Default is automatically calculated based on file size. Decimal fractions are allowed - e.g. 0.25")
 	cpCmd.PersistentFlags().StringVar(&raw.logVerbosity, "log-level", "INFO", "define the log verbosity for the log file, available levels: INFO(all requests/responses), WARNING(slow responses), ERROR(only failed requests), and NONE(no output logs).")
-	cpCmd.PersistentFlags().StringVar(&raw.blobType, "blob-type", "None", "defines the type of blob at the destination. This is used in case of upload / account to account copy")
+	cpCmd.PersistentFlags().StringVar(&raw.blobType, "blob-type", "None", "defines the type of blob at the destination. This is used in case of upload / account to account copy. Use --blob-type detect for auto-detection.")
```
[the remark this answers is not in the document]
Comment (adreed-msft) [reply cmd/copy.go]: Fair point, though, read my comment on the above conversation. (Basically using detect as a way to explicitly define detection for s2s copies, and defaulting to the source type if not explicitly defined.)

File: ste/xfer.go (hunk 2 above)
Comment (adreed-msft) [reply ste/xfer.go]: I did actually test this by hand back when I was building this. Explicit blob types are not broken, and inference does not override them.

**Comment by adreed-msft** [comment]
>Overall, I'm uneasy about the approach of using transfer.BlobType to hold the result of the inference. It's using a field which I think was meant to store "what is the blob type at the source" to now hold "what blob type have we decided to transfer this as". That change in meaning is the root cause of 3 or 4 of the issues identified in my specific comments.
>So, I suggest we change the approach, ... and instead do everything in the STE.

Honestly, I feel as though using the architecture that's already in place is the better implementation here. Yes, it changes the meaning of transfer.BlobType. However, I do feel as though transfer.BlobType is a better interface for delivering the blob type to STE rather than the JobPartPlanHeader. As is, the JPPH is monolithic, and only one is ever created.

I have a feeling we're going to have multiple blob types in one transfer at some point in the future (perhaps --list-of-files could end up supporting per-file blob type), and this is just a better way to handle it, imo.

That being said, let's at least consider the STE-only implementation.

Doing it the way you've suggested in your email as well as here would as a result, infer the blob type every time the transfer was attempted. On one hand, we're just matching against a list of strings for extensions. Not a huge deal. However, this isn't storing them in the plan files. I **think** @zezha-msft suggested that we stored the blob type in the plan files in some way. Would a different field for the destination blob type be preferable?

The other thought I have on hand is that this would at least make the copy enumerator refactor (somewhat) less of a pain, although this'd just live in the processor function anyway.

File: cmd/copyS2SMigrationBlobEnumerator.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (adreed-msft) [reply cmd/copyS2SMigrationBlobEnumerator.go]: Your answer is: We already lie to the system.

Check `ste/s2sCopier-URLToBlob.go:43`. We manually override from STE. This scenario is no different in implementation.

File: testSuite/scripts/test_autodetect_blob_type.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (adreed-msft) [reply testSuite/scripts/test_autodetect_blob_type.py]: Followed what I suggested in the latest changes, please comment if you think a different path would be optimal or resolve if this is acceptable.

File: cmd/copy.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (JohnRusk) [reply cmd/copy.go]: We can't say "Use detect for auto-detection" if auto detection also happens with None. That's misleading to users, since it implies that auto detection will not happen with anything other than "detect".

I suspect that the benefit of having None and Detect (which applies only to S2S from blob storage) is not enough to justify the complexity of having both and _explaining_ both to users.

File: testSuite/scripts/test_autodetect_blob_type.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JohnRusk) [reply testSuite/scripts/test_autodetect_blob_type.py]: As noted in a different comment today, I think supporting both becomes too hard to explain to users.

File: cmd/copyS2SMigrationBlobEnumerator.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (JohnRusk) [reply cmd/copyS2SMigrationBlobEnumerator.go]: Also, please take a look at the rest of my comment above. I'm fairly sure this PR is introducing a bug, and we can't ship with that.  Details in original comment above.  And that bug is another key reason why we should do everything in the STE. Since it's the easiest way to fix that bug. The only other way to fix that bug is to introduce a new field in the plan files (as you raised in another comment) to store _desired destination blob type_ in a new field, and keep the existing _source blob type_ as is.

[the remark this answers is not in the document]
Comment (JohnRusk) [reply cmd/copyS2SMigrationBlobEnumerator.go]: Correction: it looks like there's code in line 194 of sender-pageBlob that will actually prevent the bug from triggering. So even though in line 43 of sender-pageBlobFromURL we will actually pick up an invalid tier name, line 194 of sender-pageBlob will prevent it from being used.  So there won't be any problem at runtime.  I just confused me when reading the source, since I didn't realise we had _two_ levels of protection against the problem.

So there is no tier preservation bug.  Sorry for the false alarm.

File: cmd/copy.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (zezha-msft) [reply cmd/copy.go]: I think the 3 options can perhaps exist side by side, but with the following meanings:

- None: no detection at all, use block blob by default (for both upload and S2S).
    - For context, people may backup page blobs into block blobs to save on cost.
- Detect: smartly infer which file should be uploaded as page blob when uploading, and preserve the source type when S2S.
    - Detect may be necessary since the user may not want "smart" behavior by default, so we should let them opt-in.
- Specific blob type: override, upload all files to this kind of blobs
    - Basically telling AzCopy to shut up and just do the transfer hihi.

File: cmd/copyS2SMigrationS3Enumerator.go [hunk 7]
```diff
@@ -332,6 +332,12 @@ func (e *copyS2SMigrationS3Enumerator) addTransfersFromBucket(ctx context.Contex
 func (e *copyS2SMigrationS3Enumerator) addObjectToNTransfer(srcURL, destURL url.URL, objectInfo *minio.ObjectInfo,
 	cca *cookedCopyCmdArgs) error {
 	oie := common.ObjectInfoExtension{ObjectInfo: *objectInfo}
+	util := copyHandlerUtil{}
+
+	bt := e.BlobAttributes.BlobType.ToAzBlobType()
+	if e.BlobAttributes.BlobType == common.EBlobType.Detect() || e.BlobAttributes.BlobType == common.EBlobType.None() {
```
Comment (zezha-msft) [inline cmd/copyS2SMigrationS3Enumerator.go]: I don't think None and Detect should have equal behavior.

File: cmd/copy.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (zezha-msft) [reply cmd/copy.go]: I'm thinking that for File -> Blob and S3 -> Blob copies, if --blob-type=detect, we should infer the VHD files just like uploading. I'll confirm the expected behavior with the Stg Exp Team.

Please let me know your thoughts.

File: testSuite/scripts/test_autodetect_blob_type.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zezha-msft) [reply testSuite/scripts/test_autodetect_blob_type.py]: Good points. Let us discuss this in the comment about the flag design.

File: cmd/copy.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (JohnRusk) [reply cmd/copy.go]: Yeah, confirming with Stg Exp team sounds good. As does inference same is uploading for copying from File and S3.

[the remark this answers is not in the document]
Comment (JohnRusk) [reply cmd/copy.go]: Re the 3 possibilities of None, Detect and giving a specific type, wouldn't "None" be identical to - - blob-type BlockBlob?  Is that confusing to us or to users? Or maybe it would be more confusing, in the code, if we did _not_have None. So maybe have both None and BlockBlob, but don't document None at all???

File: cmd/copyS2SMigrationBlobEnumerator.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (adreed-msft) [reply cmd/copyS2SMigrationBlobEnumerator.go]: Ah, no worries.

File: cmd/copy.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (adreed-msft) [reply cmd/copy.go]: I mean, None should stick around. None is intended for Files and ADLSG2.

I do think that it may be a good idea, usability wise, to hide None from the user and just default to BlockBlob; switching to None for Files and ADLSG2 transfers.

File: cmd/copy.go [hunk 8]
```diff
@@ -1163,7 +1166,7 @@ func init() {
 	// options change how the transfers are performed
 	cpCmd.PersistentFlags().Float64Var(&raw.blockSizeMB, "block-size-mb", 0, "use this block size (specified in MiB) when uploading to/downloading from Azure Storage. Default is automatically calculated based on file size. Decimal fractions are allowed - e.g. 0.25")
 	cpCmd.PersistentFlags().StringVar(&raw.logVerbosity, "log-level", "INFO", "define the log verbosity for the log file, available levels: INFO(all requests/responses), WARNING(slow responses), ERROR(only failed requests), and NONE(no output logs).")
-	cpCmd.PersistentFlags().StringVar(&raw.blobType, "blob-type", "None", "defines the type of blob at the destination. This is used in case of upload / account to account copy")
+	cpCmd.PersistentFlags().StringVar(&raw.blobType, "blob-type", "Detect", "defines the type of blob at the destination. This is used in case of upload / account to account copy. Use --blob-type detect for auto-detection.")
```
[the remark this answers is not in the document]
Comment (zezha-msft) [reply cmd/copy.go]: Clarify that it detects the blob-type if there's no source blob type info available, e.g. when uploading from local, where VHD(VHDX) are assumed to be page blobs.

File: common/fe-ste-models.go [hunk 9]
```diff
@@ -395,6 +395,8 @@ func (BlobType) PageBlob() BlobType { return BlobType(2) }
 
 func (BlobType) AppendBlob() BlobType { return BlobType(3) }
 
+func (BlobType) Detect() BlobType { return BlobType(4) }
```
Comment (zezha-msft) [inline common/fe-ste-models.go]: this is replacing None as the default value, is None ever used? If not, it can be removed, right?

Comment (JohnRusk) [reply common/fe-ste-models.go]: If we remove None, then should we make Detect have the value zero?  (Given that zero-like values are kind of special in Go, as the default for new variables.  That's one thing that's nice about None, in that it's name implies zero-like-ness. But I agree that from a user's viewpoint its redundant)

Comment (zezha-msft) [reply common/fe-ste-models.go]: Indeed, the default value can be straight up `Detect`.

**Comment by zezha-msft** [comment]
@adreed-msft :shipit:
